block/src/Pos/Block/Logic/Header.hs

```diff
@@ -352,29 +357,39 @@ getHeadersOlderExp upto = do
                 | otherwise = selGo es ii $ succ skipped
         in selGo elems ixs 0
 
+
+data GetHashesRangeException = GetHashesRangeException Text deriving (Show,Generic)
+instance Exception GetHashesRangeException
+
+-- | Throws 'GetHashesRangeException'.
+throwGHR :: MonadThrow m => Text -> m a
+throwGHR = throwM . GetHashesRangeException
+
 -- | Given optional @depthLimit@, @from@ and @to@ headers where @from@
 -- is older (not strict) than @to@, and valid chain in between can be
 -- found, headers in range @[from..to]@ will be found. If the number
 -- of headers in the chain (which should be returned) is more than
 -- @depthLimit@, error will be thrown.
-getHeadersRange ::
+getHashesRange ::
        forall m. (HasConfiguration, MonadDBRead m)
     => Maybe Word
     -> HeaderHash
     -> HeaderHash
-    -> m (Either Text (OldestFirst NE HeaderHash))
-getHeadersRange depthLimitM older newer | older == newer = runExceptT $ do
+    -> m (OldestFirst NE HeaderHash)
+getHashesRange depthLimitM older newer | older == newer = do
     unlessM (isJust <$> DB.getHeader newer) $
-        throwError "getHeadersRange: can't find newer-older header"
+        throwGHR "can't find newer-older header"
     whenJust depthLimitM $ \depthLimit ->
         when (depthLimit < 1) $
-        throwError $
-        sformat ("getHeadersRange: depthLimit is "%int%
+        throwGHR $
```

Comment:
Again, really an error? Can't we handle it non-exceptionally?

Reply:
Again, `throwError` that was there before is not "non-exceptionally" either, we should either use `throwM` here or a non-monadic way to handle this codepath.